### PR TITLE
Json compare

### PR DESCRIPTION
**Pull request by Christopher-Bradshaw**
This adds the option to do json diffing. We try a simple string match between the bodies and if that fails we try to do a json comparison.

In the json comparison, we ignore differences in the order of keys (correct according to the json spec), and add the option (by default OFF) to ignore ordering in arrays (incorrect according to the json spec). However we need to do this second part as in some places we use go maps.

These is a bit of refactoring to allow us to do this.

Fairly extensive tests of the comparisons have been added.

I'll run some sanity tests in dev to test this too.

### CODE REVIEW

**Comment by kvigen** [comment]
Hmm, that's a good question. Ordering in arrays matters for anything that has sorting / paging, and that seems important.

I'm not sure how often we turn maps into arrays. That we do it for scopes in district-auth is more for legacy reasons than because we think it's a good idea.

My gut would be to fix the district-auth issue by sorting the scopes there, and only add the strong / weak distinction if we keep having to do that. Last year when we were testing the sis-api with science we added implicit `_id` sorting to a few endpoints to solve similar problems, and I think it was a good idea (though it's a bit different from the district-auth case)

**Comment by Christopher-Bradshaw** [comment]
Yeah, I probably should have thought about **why** I was doing this before doing it... Anyway

I'm going to think about this a bit more but think I will end up with a simple json diff `reflect.deepEqual()` (I'll make sure that works how we want it to) by default, and then have this as an option if you specify something in the payload.

**Comment by Christopher-Bradshaw** [comment]
This is almost ready to go. One pathological case (duplicates in an array) that I need to deal with.

File: .drone.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ publish:
     - $(git rev-parse --short HEAD)
     username: $$docker_username
     when:
-      branch: master
+      branch: json-compare
```
Comment (Christopher-Bradshaw) [inline .drone.yml]: REVERT

**Comment by Christopher-Bradshaw** [comment]
cc @kvigen I think this is ready if you also want to take a look

File: science/common.go [hunk 2]
```diff
@@ -21,38 +22,51 @@ type Results struct {
...
 	if err = r.Write(conn); err != nil {
-		return "", 0, fmt.Errorf("error writing request to %s: %s", addr, err)
+		return &forwardedRequest{}, fmt.Errorf("error writing request to %s: %s", addr, err)
 	}
 	res, err := http.ReadResponse(bufio.NewReader(conn), r)
 	if err != nil {
-		return "", 0, fmt.Errorf("error reading response from %s: %s", addr, err)
+		return &forwardedRequest{}, fmt.Errorf("error reading response from %s: %s", addr, err)
 	}
-	defer res.Body.Close()
```
Comment (meliaj) [inline science/common.go]: I think we still want this somewhere

File: science/compare.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	if len(a) != len(b) {
+		return false
+	}
+	used := []int{}
+
+	for _, v1 := range a {
+		areEqual := false
+		for i, v2 := range b {
+			// Check that we have't already used this element to match
+			if inList(i, used) {
```
Comment (meliaj) [inline science/compare.go]: `i` is the index. How does this work?

Comment (Christopher-Bradshaw) [inline science/compare.go]: If we have used this index, we can't use it again.

I could maybe make this clearer by changing names.

Comment (Christopher-Bradshaw) [inline science/compare.go]: See `used = append(used, i)` in 129

Comment (meliaj) [inline science/compare.go]: Got it 👍

**Comment by meliaj** [comment]
One nit to close the response body but otherwise LGTM

File: science/common.go (hunk 2 above)
Comment (Christopher-Bradshaw) [inline science/common.go]: 68fd3bc 👍

**Comment by Christopher-Bradshaw** [comment]
Ran in dev with AVS and an old version of AVS. See https://cb-ctrl--gearman-admin.int.clever.com/jobs/4, and results for posterity: `Results map[int]map[int]int{200:map[int]int{200:333}, 500:map[int]int{200:2255, 404:9}},`

Note that these were mostly diffs in body, not in headers/codes which is what the changes here would effect.

**Pull request merged**